crates/ide-assists/src/handlers/desugar_doc_comment.rs

```diff
@@ -0,0 +1,313 @@
+use either::Either;
+use itertools::Itertools;
+use syntax::{
+    ast::{self, edit::IndentLevel, CommentPlacement, Whitespace},
+    AstToken, TextRange,
+};
+
+use crate::{AssistContext, AssistId, AssistKind, Assists};
+
+use super::{
+    convert_comment_block::{line_comment_text, relevant_line_comments},
+    raw_string::required_hashes,
+};
+
+// Assist: desugar_doc_comment
+//
+// Desugars doc-comments to the attribute form.
+//
+// ```
+// /// Multi-line$0
+// /// comment
+// ```
+// ->
+// ```
+// #[doc = r"Multi-line
+// comment"]
+// ```
+pub(crate) fn desugar_doc_comment(acc: &mut Assists, ctx: &AssistContext<'_>) -> Option<()> {
+    let comment = ctx.find_token_at_offset::<ast::Comment>()?;
+    // Only allow doc comments
+    let Some(placement) = comment.kind().doc else { return None; };
+
+    // Only allow comments which are alone on their line
+    if let Some(prev) = comment.syntax().prev_token() {
+        if Whitespace::cast(prev).filter(|w| w.text().contains('\n')).is_none() {
+            return None;
+        }
+    }
+
+    let indentation = IndentLevel::from_token(comment.syntax()).to_string();
+
+    let (target, comments) = match comment.kind().shape {
+        ast::CommentShape::Block => (comment.syntax().text_range(), Either::Left(comment)),
+        ast::CommentShape::Line => {
+            // Find all the comments we'll be desugaring
+            let comments = relevant_line_comments(&comment);
+
+            // Establish the target of our edit based on the comments we found
+            (
+                TextRange::new(
+                    comments[0].syntax().text_range().start(),
+                    comments.last().unwrap().syntax().text_range().end(),
+                ),
+                Either::Right(comments),
+            )
+        }
+    };
```

Comment:
This can also be moved into the builder closure

Reply:
well we need the `target` for `acc.add`

Reply:
We could move the indentation inside the body, but then it would look like this:
```rs
            let indentation = comments
                .as_ref()
                .map_left(|c| IndentLevel::from_token(c.syntax()))
                .left_or_else(|comments| IndentLevel::from_token(comments[0].syntax()))
                .to_string();
```

Reply:
And while we could only collect the target but not the comments and recollect them inside the closure, that didn't make a lot of sense to me.